src/hotspot/cpu/x86/macroAssembler_x86_adler.cpp

```diff
@@ -0,0 +1,211 @@
+*
+* This code is free software; you can redistribute it and/or modify it
+* under the terms of the GNU General Public License version 2 only, as
+* published by the Free Software Foundation.
+*
+* This code is distributed in the hope that it will be useful, but WITHOUT
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+* version 2 for more details (a copy is included in the LICENSE file that
+* accompanied this code).
+*
+* You should have received a copy of the GNU General Public License version
+* 2 along with this work; if not, write to the Free Software Foundation,
+* Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+*
+* Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+* or visit www.oracle.com if you need additional information or have any
+* questions.
+*
+*/
+
+#include "precompiled.hpp"
+#include "asm/assembler.hpp"
+#include "asm/assembler.inline.hpp"
+#include "runtime/stubRoutines.hpp"
+#include "macroAssembler_x86.hpp"
+
+#ifdef _LP64
+void MacroAssembler::updateBytesAdler32(Register init_d, Register data, Register size, XMMRegister yshuf0, XMMRegister yshuf1, ExternalAddress ascaletab)
+{
+      const int LIMIT = 5552;
+      const int BASE = 65521;
+      const int CHUNKSIZE =  16;
+      const int CHUNKSIZE_M1 = CHUNKSIZE - 1;
+
+      const Register s = r11;
+      const Register a_d = r12; //r12d
+      const Register b_d = r8; //r8d
+      const Register end = r13;
+
+      const XMMRegister ya = xmm0;
+      const XMMRegister yb = xmm1;
+      const XMMRegister ydata0 = xmm2;
+      const XMMRegister ydata1 = xmm3;
+      const XMMRegister ysa = xmm4;
+      const XMMRegister ydata = ysa;
+      const XMMRegister ytmp0 = ydata0;
+      const XMMRegister ytmp1 = ydata1;
+      const XMMRegister ytmp2 = xmm5;
+      const XMMRegister xa = xmm0;
+      const XMMRegister xb = xmm1;
+      const XMMRegister xtmp0 = xmm2;
+      const XMMRegister xtmp1 = xmm3;
+      const XMMRegister xsa = xmm4;
+      const XMMRegister xtmp2 = xmm5;
+      assert_different_registers(init_d, data, size, s, a_d, b_d, end, rax);
+
+      Label SLOOP1, SLOOP1A, SKIP_LOOP_1A, FINISH, LT64, DO_FINAL, FINAL_LOOP, ZERO_SIZE, END;
+
+      push(r12);
+      push(r13);
+      push(r14);
+      movl(b_d, init_d); //adler
+      shrl(b_d, 16);
+      andl(init_d, 0xFFFF);
+      cmpl(size, 32);
+      jcc(Assembler::below, LT64);
+      movdl(xa, init_d); //vmovd - 32bit
+      vpxor(yb, yb, yb, Assembler::AVX_256bit);
+
+      bind(SLOOP1);
+      movl(s, LIMIT);
+      cmpl(s, size);
+      cmovl(Assembler::above, s, size); // s = min(size, LIMIT)
+      lea(end, Address(s, data, Address::times_1, -CHUNKSIZE_M1));
+      cmpptr(data, end);
+      jcc(Assembler::aboveEqual, SKIP_LOOP_1A);
+
+      align(32);
+      bind(SLOOP1A);
+      vbroadcastf128(ydata, Address(data, 0), Assembler::AVX_256bit);
```

Comment:
Loop operates over integral data (double word), shouldn't it be safe to use double broadcasting instruction to save which domain switchover penalty (not sure if broadcasting will causes domain switch over though), but this is happening in the main vector loop so being conservative.